src/platform/i386/chal/chal_plat.h

```diff
@@ -7,24 +7,76 @@ int chal_tlbstall(void);
 int chal_tlbstall_recount(int a);
 int chal_tlbflush(int a);
 
+
+#if defined(__x86_64__)
+#define CR3_NO_FLUSH (1ul << 63)
+#endif
+
+/* Reference Intel 64 and IA-32 Architecture Software Developer's Manual, Volume 2 */
+#define INVPCID_TYPE_INDIVIDUAL_ADDR    0 /* invalidate all tlb entries for a PCID used to map a vaddr */
+#define INVPCID_TYPE_SINGLE_CONTEXT     1 /* invalidate all tlb entries (not global) for a PCID */
+#define INVPCID_TYPE_ALL_CONTEXT_GLOBAL 2 /* invalidate all tlb entries, including global */
+#define INVPCID_TYPE_ALL_CONTEXT        3 /* invalidate all tlb entries, not including global */
+
+static inline void
+__invpcid(u64_t pcid, u64_t addr, unsigned long type)
+{
+	/* 
+	 * invpcid takes a 128 bit value from memory as: | address | 000...0 | pcid | 
+	 *                                              127        63        11     0
+	 * This is the case in both 32 and 64 bit execution modes
+	 */ 
+	struct { u64_t pcid; u64_t addr; } desc = { .pcid = pcid, .addr = addr };
+
+	asm volatile("invpcid %0, %1" : : "m"(desc), "r"(type) : "memory");
+}
+
+static inline unsigned long
```

Comment:
I don't see where this and __write_cr3 are used. __write* I can see being used to switch page-tables. but __readcr3?